### ISSUE TITLE
Difficulty defining relationships

### ISSUE THREAD

**Issue by Aaronfischer** [samselikoff/ember-cli-mirage#687]
I've read that one-to-one relationships are not directly supported, so then my next question is, what kind of an impact does this actually have?
Am I still able to load `/assets/` that has a one-to-one belongsTo relationship with `/assets/:id/images` and have it fetch the belongsTo `image`?

I'm having a tough time configuring this setup, heres what I have so far:

``` js
// app/models/asset.js
export default DS.Model.extend({
    image: DS.belongsTo('asset/image', {async: true}),
    title: attr(),
});

// app/models/asset/image.js
export default DS.Model.extend({
    asset: DS.belongsTo('asset', {async: true}),
    title: attr(),
});

// mirage/factories/asset.js
export default Mirage.Factory.extend({
    imageId(i) { return i + 1; },

    title() { return faker.name.firstName() + ' ' + faker.name.lastName(); },
});

// mirage/factories/image.js
export default Mirage.Factory.extend({
    title() { return faker.name.firstName() + ' ' + faker.name.lastName(); },
    createdAt() { return moment(faker.date.past()).format('YYYY-MM-DD'); },
    updatedAt() { return moment(faker.date.past()).format('YYYY-MM-DD'); },
});

// mirage/models/asset.js
export default Model.extend({
    image: belongsTo('image'),
});

// mirage/models/image.js
export default Model.extend({
    asset: belongsTo('asset'),
});

// mirage/scenarios/default.js
export default function(server) {
  server.createList('image', 10);
  server.createList('asset', 10);
}

// mirage/applications/asset.js
/* NOTE: I'm not sure if I can use "relationship:" to have it lazyload the data or not? */
// ApplicationSerializer is JSONAPIserializer
export default ApplicationSerializer.extend({
    include: ['image'],
});

// mirage.config.js
...
    /**
     * Assets
     */
    this.get('/assets', (schema, request) => {
        if (request.queryParams) {
            return schema.asset.where(Utils.filterObj(request.queryParams, schema.db['assets'], {}));
        }

        return schema.asset.all();
    });

    this.get('/assets/:id/images', (schema, request) => {
        let asset = schema.asset.find(request.params.id);

        return asset.images;
    });
}

```

The response from /assets looks exactly like it should, its an array of objects and each object has a `relationships.image.data` which is `{id: '1', type: 'images'}`.

However, the /assets/:id/images are not loading and no request is being made for them anywhere. I'm sure I could have missed something in my setup, do you see any mistakes?
I'd love to share a link to the project, but I'm unable to (private repo). Let me know if there is anything else you need, other than this issue I've loved switching over to mirage, brilliantly crafted!

**Comment by samselikoff** [comment]
First question, are you saying your Ember app never makes a request for the assets' images at all? In that case, either there's a problem with the `/assets` response, or your Ember app isn't doing what you expect (meaning, it's an Ember problem not a Mirage problem).

Usually with async you would send up a `link` property, which Ember Data would then follow to get the assets' links.

Second, in `scenarios/default` you create 10 images and assets, and your asset factory specifies a `imageId` attribute. Instead you should associate the keys in the scenario itself:

```js
// mirage/scenarios/default.js
export default function(server) {
  for (var i = 0; i < 10; i++) {
    let image = server.create('image');
    server.create('asset', { image })
  }
}
```

Basically, you don't want to "guess" at the foreign keys, you want to use them after you've created the records.

Does that all make sense?

**Comment by samselikoff** [comment]
In your actual backend, what is the relationship? `asset` `has_one` `image`?

**Comment by Aaronfischer** [comment]
Hey @samselikoff, thanks!

1. It looks like there was a problem with some other serializing I was doing that was not taking into account the `included` object (so if my model serializer had "include: image" then it looks like its included in the response)
  * I can correct for this, this also brings about an issue with ember expecting a model at `asset/image`, but its returning `image`, is there a way to setup this "nested" idea in mirage?
  * Perhaps this is why a request isn't being made then because of this mismatch?
2. So this is something that needs to happen in the scenarios file then? So `imageId` defined in the factory isn't necessary unless we aren't associating them within the scenario.js?
3. `asset` belongsTo `image` and `image` belongsTo `asset`

We use the `links.related` for hasMany relationships typically, but is mirage expecting this `link` property for any relationship type? So is this something I need to implement in addition to the above?

**Comment by Aaronfischer** [comment]
Also, you are correct, if I was to remove `include: image`, no request is made to fetch (or lazyload) the asset's image.

**Comment by samselikoff** [comment]
There's a lot going on here! I think it would be easiest if you could show me the response you're trying to mock out, one at a time. Then we can work from there.

**Comment by Aaronfischer** [comment]
Yes, there is! The initial response on /assets would look like:
```js
// /assets
"data": [
    {
        "id": "1",
        "type": "assets",
        "attributes": {
            "title": "Asset Title",
        },
        "relationships": {
            "image": {
                "data": {"type": "images", "id": "1"}
            },
            "images": {
                "links": {
                    "related": "/assets/1/images"
                }
            }
        }
    }
    ...
]
```

Then I'd expect the asset image to be lazy-loaded and a request would be made to `/assets/:id/images/:id` (this would be to grab the `image` relationship)
```js
// assets/:id/images/:id
"data": {
    "id": "1",
    "type": "images",
    "attributes": {
        "title": "Asset Image Title",
        "order": 1,
    },
    "relationships": {
        "asset": {
            "data": {"type": "assets", "id": "1"}
        },
        "file": {
            "data": {"type": "files", "id": "1"}
        }
    }
}
```

There would then be an additional request made after this to grab the actual url/etc via `file`, but one step at a time. :+1: 

Does this help clear things up?

**Comment by samselikoff** [comment]
Does `asset` both belongTo and hasMany images? Or, why are there two keys under `relationships` related to `image`?

**Comment by Aaronfischer** [comment]
Yep, it would be used primarily as a way to get the primary "image" verse having to iterate over the array of "images" just to grab the first(primary) one.

I'm just at a bit of a loss on how to define these relationships in mirage. Thanks!

**Comment by samselikoff** [comment]
I think this points to a need for a new API on the serializers, to specify how relationship references are serialized (links vs. data). Currently, references use `data` so the `image` relationship should be handled automatically. For the `images` `links` bit, for now I'd just do that manually in the serializer.

[I've made a twiddle here](https://canary.ember-twiddle.com/523f22c8104d702d48b5fcc9be7ad583?openFiles=mirage.serializers.asset.js%2C). The asset serialize looks like this:

```js
import ApplicationSerializer from './application';

export default ApplicationSerializer.extend({
  
  serialize(response, request) {
    let json = ApplicationSerializer.prototype.serialize.apply(this, arguments);
    
    json.data.forEach(object => {
      object.relationships.images = {
        related: `/assets/${object.id}/images`
      };
    });
    
    return json;
  }
  
});
```

See if that gets you started and let me know if you have more questions. Thanks for opening this issue, I know the API still has a way to go before this kind of stuff is painless!

**Comment by Aaronfischer** [comment]
Thanks for taking the time to setup an example twiddle, this really helped!
I'm definitely starting to understand how to configure/define a flat relationship, however, how would the above "nested" relationship be defined in mirage, perhaps this is where I'm missing a step. 

When configuring nested relationships (ex. /assets/:id/images) does/should mirage follow the same nested structure as Ember? I've forked your twiddle and have attempted to setup the nested `asset/image` structure on the Ember side and am then attempting to iterate over `asset.images` to return each image's title just as a basic example.
https://canary.ember-twiddle.com/48847ca5f2a0391ee3abbab6e45bfd05?openFiles=templates.application.hbs%2C

(Getting a few errors, but I think you can see from my setup/changes what my attempts are at getting a nested structure.)

**Comment by samselikoff** [comment]
Right now Mirage doesn't have any built-in support for nested model names. We can munge this a bit in the serializer, but my first question is, is this how your current server handles it, or are you building this out in a greenfield app? If the latter, you don't need to nest the model name `asset/image` just to have the related `/assets/:id/images` endpoint. The model could still be just `image` (and in this case, I think your twiddle fork would work).

**Comment by samselikoff** [comment]
If you want to keep the nested model name (`asset/image`), I think you need to change the `type` property on the `image` Resource Object in the json:api payload:

![image](https://cloud.githubusercontent.com/assets/2922250/15048002/76e4291c-12b5-11e6-8709-f573e407de59.png)

Ember doesn't have any model called `image`, it has one called `asset/image`, so you need this `type` property to be `asset/image`. You could do that by overwriting `serialize` on the `image-serializer`.
